Minmod performance improvements

nilsdeppe opened the pull request:
## Proposed changes

- Add `gsl::span`, which provides a way to get a view on data.
- Reduce allocations
- Reduce number of times `SliceIterator` is used

### Types of changes:

- [ ] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add mean_value_on_boundary overload with boundary_buffer

- Allows us to reuse a boundary buffer for repeated calls to
  `mean_value_on_boundary`
- Adds an overload that takes volume and slice indices to avoid calling the
  slice iterator.

nilsdeppe committed:
Use a temp buffer in Minmod for mean_value_on_boundary

nilsdeppe committed:
Add gsl::span class

nilsdeppe committed:
Reduce test cases for mean_value

moxcodes commented on src/Utilities/Gsl.hpp:
```diff
@@ -266,6 +274,585 @@ not_null<T> operator+(const not_null<T>&, std::ptrdiff_t) = delete;
+}
+
+#if __GNUC__ > 6
+#pragma GCC diagnostic push
```
`GCC diagnostic pop`

moxcodes commented on src/Utilities/Gsl.hpp:
```diff
@@ -266,6 +274,585 @@ not_null<T> operator+(const not_null<T>&, std::ptrdiff_t) = delete;
+/// \ingroup UtilitiesGroup
+/// Utility function for creating spans
+template <class ElementType>
+constexpr span<ElementType> make_span(
```
There are several functions in this commit that are not marked noexcept. I understand this is likely because the gsl versions are probably not marked noexcept for good reason, but it also seems our version doesn't throw, so I'm just curious of our policy for this type of edge case.

nilsdeppe replied:
I actually don't know. I'd be fine with marking them `noexcept`. @wthrowe @kidder thoughts?

kidder replied:
I don't care

wthrowe replied:
There's a lot of stuff in this file that could be noexcept but isn't, so let's leave this as is and we can go through and change the whole file later if we want to.

moxcodes commented on src/NumericalAlgorithms/LinearOperators/MeanValue.hpp:
```diff
@@ -46,11 +49,42 @@ double mean_value(const DataVector& f, const Mesh<Dim>& mesh) noexcept {
+ * - `boundary_buffer` is a pointer to a DataVector of size
+ *   `mesh.slice_away(d).number_of_grid_points()` used as a temporary buffer
+ *   when slicing the data to the boundary. In the 1D case no buffer is required
+ *   and a `nullptr` may be passed.
```
As the specialization already exists, is it more elegant to include a default for the unused variable?

nilsdeppe replied:
You can't default arguments that aren't the last ones (as in, you could default the last 5, but not the first 1)

moxcodes commented on src/NumericalAlgorithms/LinearOperators/MeanValue.cpp:
```diff
@@ -10,29 +10,86 @@
+template <size_t Dim>
+double mean_value_on_boundary(const DataVector& f, const Mesh<Dim>& mesh,
+                              size_t d, Side side) noexcept {
+  ASSERT(d < Dim, "d = " << d << ", Dim = " << Dim);
```
Would this `ASSERT` be better placed in the `not_null` version, so it is present for all execution paths?

moxcodes commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Minmod.hpp:
```diff
@@ -366,26 +371,70 @@ class Minmod<VolumeDim, tmpl::list<Tags...>> {
+                         gsl::span<std::pair<size_t, size_t>>>,
+               VolumeDim>
+        volume_and_slice_indices{};
+    make_overloader(
```
for my own education, why is this better than an if else on `VolumeDim`?

nilsdeppe replied:
Good question! With an if statement both branches need to be compiled. However, the 1D case does not work because SliceIterator (or maybe a different class, I can't remember which one) doesn't exist in 1-1(=0)D. The second lambda is only instantiated if the first lambda is not a match, thus avoiding the 1D case. The same can be achieved in C++17 with `if constexpr`. Does that clarify?

moxcodes replied:
it does; thanks!

moxcodes reviewed (changes requested):
A small pragma typo, and a few curiosities.

nilsdeppe commented:
Addressed what I could, waiting for further discussion on the rest :)

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Minmod.hpp:
```diff
@@ -355,10 +362,79 @@ class Minmod<VolumeDim, tmpl::list<Tags...>> {
+                                           num_points);
+      alloc_offset += num_points;
+    }
+    // Compute the slice indices once since this is (surprisingly) expensive
```
This computation and the one in the `mean_value_on_boundary` test share a lot of code, and I suspect any future users of `mean_value_on_boundary` will need to reimplement this once again. Could you make it a free function so we can reduce the duplication?

nilsdeppe replied:
I agree, it'll be used a lot. You're asking for a function that would return a `std::pair<unique_ptr, std::array<...>>`?

fmahebert replied:
I'm not sure. I was roughly thinking of a function that would compute and return `volume_and_slice_indices` (as declared in line 393 below). Perhaps the user would still have to allocate the `indices_buffer` and pass it into the function... which would be okay, though cleaner if avoidable.

Edit: I'm not sure how the return type you suggested fits in.

nilsdeppe replied:
Ah, the return type comes in because the caller of the function needs to take ownership of the memory as well as the views into that memory. The unique_ptr holds the memory, the array holds the views. Does that make sense?

fmahebert replied:
Ah, so the function would return a pair containing `indices_buffer` and `volume_and_slice_indices`? If so, then yes this is what I was asking for :+1:

nilsdeppe replied:
Yes, that's what I had in mind. I'll factor it out :+1: I'm not sure where to put the function. Right now "MeanValue" seems reasonable, but so would `SliceIterator` or a new header :/

fmahebert replied:
Great! To me `SliceIterator` or a new header seems more natural than with `MeanValue`... ¯\\\_(ツ)\_/¯

kidder replied:
I'd put it in `SliceIterator`

kidder replied:
what is the status of this?

kidder commented on src/Utilities/Gsl.hpp:
```diff
@@ -266,6 +274,585 @@ not_null<T> operator+(const not_null<T>&, std::ptrdiff_t) = delete;
 
+// GCC 7 does not like the signed unsigned missmatch (size_t ptrdiff_t)
+// While there is a conversion from signed to unsigned, it happens at
+// compiletime, so the compiler wouldn't have to warn indiscriminently, but
```
compile time

kidder replied:
ping

wthrowe commented on src/NumericalAlgorithms/LinearOperators/MeanValue.cpp:
```diff
@@ -10,29 +10,86 @@
 
-template <>
-double mean_value_on_boundary(const DataVector& f, const Mesh<1>& mesh,
+double mean_value_on_boundary(const DataVector* /*unused*/,
```
Shouldn't this be a not_null to match the other versions?  This could be annoying in generic code where you already have a not_null buffer.

It seems like the null-pointer benefit is only going to be useful if you know you're in 1D, but then you could just use the no-scratch version of the function.  And empty DataVectors are cheap anyway.

nilsdeppe replied:
Agreed :)

wthrowe commented:
Looks good.  Squash.

nilsdeppe commented:
Okay, squashed!

nilsdeppe committed:
Add function to provide volume and slice indices

nilsdeppe committed:
Make libsharp CMake message consistent

nilsdeppe commented:
- Rebased on develop
- Added the `SliceIterator` function in a separate commit
- Fixed libsharp CMake messages to be consistent with other libs

kidder reviewed (dismissed):
@wthrowe @fmahebert please look at

wthrowe commented on src/NumericalAlgorithms/LinearOperators/MeanValue.hpp:
```diff
@@ -46,11 +49,42 @@ double mean_value(const DataVector& f, const Mesh<Dim>& mesh) noexcept {
+ * - `boundary_buffer` is a pointer to a DataVector of size
+ *   `mesh.slice_away(d).number_of_grid_points()` used as a temporary buffer
+ *   when slicing the data to the boundary. In the 1D case no buffer is required
+ *   and a `nullptr` may be passed.
```
Update this to match the code changes.

kidder commented:
@wthrowe @fmahebert please look at

fmahebert commented on src/NumericalAlgorithms/LinearOperators/MeanValue.cpp:
```diff
@@ -9,30 +9,83 @@
 #include "DataStructures/SliceIterator.hpp"
```
This commit message ("Add mean_value_on_boundary overload with boundary_buffer") should also mention the slice-to-volume index optimization.

fmahebert commented on src/NumericalAlgorithms/LinearOperators/MeanValue.cpp:
```diff
@@ -9,30 +9,83 @@
 
-template <>
-double mean_value_on_boundary(const DataVector& f, const Mesh<1>& mesh,
+double mean_value_on_boundary(const gsl::not_null<DataVector*> /*unused*/,
```
I can infer the first argument is unused because it is commented out ... how about calling it `/*boundary_buffer*/` to facilitate comparison with the general (Dim != 1) case?

fmahebert commented on src/NumericalAlgorithms/LinearOperators/MeanValue.cpp:
```diff
@@ -9,30 +9,83 @@
+                              const DataVector& f, const Mesh<1>& mesh,
                               size_t d, Side side) noexcept {
   ASSERT(d == 0, "d = " << d);
   return Side::Lower == side ? f[0] : f[mesh.extents(0) - 1];
```
Why is the 1D case implemented 3 times, instead of forwarding to one case? This is not a change request, I'm just trying to understand if you had a reason to avoid forwarding (perhaps avoiding the cost of the function call?).

fmahebert commented on src/NumericalAlgorithms/LinearOperators/MeanValue.cpp:
```diff
@@ -9,30 +9,83 @@
+    const gsl::not_null<DataVector*> boundary_buffer,
+    const gsl::span<std::pair<size_t, size_t>> volume_and_slice_indices,
+    const DataVector& f, const Mesh<Dim>& mesh, const size_t d,
+    const Side /*side*/) noexcept {
```
Add assertion on Dim here also (or, if not, why not? I care more about consistency between implementations than about the assertion)

fmahebert commented on tests/Unit/NumericalAlgorithms/LinearOperators/Test_MeanValue.cpp:
```diff
@@ -81,17 +86,79 @@ void test_mean_value_on_boundary() {
+            gsl::at(volume_and_slice_indices, d).second[index].second =
+                sli_upper.slice_offset();
+          }
+        }
```
Can't you use the generic function here when computing these indices?

fmahebert commented on tests/Unit/NumericalAlgorithms/LinearOperators/Test_MeanValue.cpp:
```diff
@@ -81,17 +86,79 @@ void test_mean_value_on_boundary() {
+        CHECK(0.0 == approx(mean_value_on_boundary(
+                         &temp_buffers[0], volume_and_slice_indices[0].second,
+                         u_quad, mesh, 0, Side::Upper)));
+
```
In the x case you interleave the "generic" and "optimized" versions of `mean_value_on_boundary`. In the y,z cases you put generic first and optimized second. Make a consistent choice (personally, I have slight preference for the way you did y,z but I don't really case).

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Minmod.hpp:
```diff
@@ -355,10 +357,35 @@ class Minmod<VolumeDim, tmpl::list<Tags...>> {
+    // Allocate temporary buffer to be used in `limit_one_tensor` where we
+    // otherwise make 1 + 2 * VolumeDim allocations per tensor component for
+    // MUSCL and Pi1, and 1 + 4 * VolumeDim allocations per tensor component for
+    // PiN.
```
I know I will sometimes say "Pi1" and "PiN" informally, but let's keep the documentation saying "LambdaPi1" and "LambdaPiN".

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Minmod.hpp:
```diff
@@ -355,10 +357,35 @@ class Minmod<VolumeDim, tmpl::list<Tags...>> {
+    // otherwise make 1 + 2 * VolumeDim allocations per tensor component for
+    // MUSCL and Pi1, and 1 + 4 * VolumeDim allocations per tensor component for
+    // PiN.
+    size_t total_boundary_points = 0;
```
If I correctly understand the code, this is just _half_ of the number of boundary points: it counts the points in one of the two constant-x slices, one of two constant-y slices, one of two constant-z slices. I agree this is correct, but the variable name should be "half_number_boundary_points" or similar.

Also you can make this const by assigning with a lambda.

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Minmod.hpp:
```diff
@@ -366,26 +370,28 @@ class Minmod<VolumeDim, tmpl::list<Tags...>> {
-                   (mesh.number_of_grid_points() + total_boundary_points))),
+        static_cast<double*>(malloc(
+            sizeof(double) * (mesh.number_of_grid_points() +
+                              (VolumeDim > 0 ? total_boundary_points : 0)))),
```
- This diff looks like it is in the wrong commit
- Here, are you avoiding allocating the extra double for the boundary of a 1D interval? If so, please add a short comment.

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Minmod.hpp:
```diff
@@ -366,26 +370,28 @@ class Minmod<VolumeDim, tmpl::list<Tags...>> {
     alloc_offset += mesh.number_of_grid_points();
     std::array<DataVector, VolumeDim> temp_boundary_buffer{};
-    for (size_t d = 0; d < VolumeDim; ++d) {
+    for (size_t d = 0; VolumeDim > 0 and d < VolumeDim; ++d) {
```
Also looks to be in wrong commit

fmahebert commented on src/DataStructures/SliceIterator.hpp:
```diff
@@ -53,3 +61,20 @@ class SliceIterator {
+
+/*!
+ * \ingroup DataStructuresGroup
+ * \brief Get the volume and slice indices.
```
I suspect this isn't very clear to someone who isn't familiar with this function already.

How about: "Get the mapping between volume and boundary slice indices" or similar?

fmahebert commented on src/DataStructures/SliceIterator.hpp:
```diff
@@ -53,3 +61,20 @@ class SliceIterator {
+ * the array holds views into the memory buffer. The index of the array is the
+ * fixed dimension, the outer `pair` holds the indices for the lower and upper
+ * side,  respectively, while the `pair`s in the `span`s hold the volume and
+ * slice indices, respectively.
```
Before the paragraph describing how the data is arranged, it would be helpful to have a 1-2 sentence paragraph explaining what this data represents, and why it is useful to compute this (i.e., optimizing SliceIterator because it is surprisingly slow).

fmahebert commented on tests/Unit/DataStructures/Test_SliceIterator.cpp:
```diff
@@ -27,18 +28,53 @@ void check_slice_iterator_helper(SliceIterator si) {
+    check_slice_and_volume_indices(Index<2>{3, 4});
+    check_slice_and_volume_indices(Index<2>{5, 4});
+    check_slice_and_volume_indices(Index<2>{2, 2});
+    check_slice_and_volume_indices(Index<1>{2});
```
Mild preference for organizing these from 1D -> 3D and "numerically" in each dimension, but I invite you to ignore my mild preference on this.

fmahebert commented on src/DataStructures/SliceIterator.cpp:
```diff
@@ -37,11 +39,74 @@ void SliceIterator::reset() {
+                               gsl::span<std::pair<size_t, size_t>>>,
+                     VolumeDim>>
+volume_and_slice_indices(const Index<VolumeDim>& extents) noexcept {
+  // Compute the slice indices once since this is (surprisingly) expensive
```
This comment doesn't need to be in the implementation if the header documents this idea.

fmahebert commented on src/DataStructures/SliceIterator.cpp:
```diff
@@ -37,11 +39,74 @@ void SliceIterator::reset() {
+             VolumeDim>
+      volume_and_slice_indices{};
+
+  size_t total_boundary_points = 0;
```
As commented elsewhere, I think this should be named "half_number_boundary_points", and it can be made const if assigned by a lambda.

nilsdeppe committed:
Reduce SliceIterator usage in Minmod

nilsdeppe committed:
Add iota that returns filled container

nilsdeppe commented:
Okay, I pushed fixup commits, rebased on develop and added a new commit that adds an `alg::iota` where you pass a container and get the container back. The goal of it is to make it easier to do these short loops over dim and return const things.

fmahebert commented on src/DataStructures/SliceIterator.hpp:
```diff
@@ -64,7 +64,14 @@ class SliceIterator {
+ * volume data and the index in the corresponding sliced data. Repeatedly
+ * applying the SliceIterator on various components of a tensor becomes very
+ * expensive and so precomputing the index map is sometimes advantageous. This
+ * functions computes the index map onto all boundary slices of volume mesh with
```
I like this comment 👍, but fix the typo: functions -> function

fmahebert commented on src/NumericalAlgorithms/LinearOperators/MeanValue.cpp:
```diff
@@ -17,22 +18,73 @@ double mean_value_on_boundary(const DataVector& f, const Mesh<1>& mesh,
+}
+
+double mean_value_on_boundary(
+    const gsl::not_null<DataVector*> /*unused*/,
```
unused -> boundary_buffer, also check for this in header file

fmahebert reviewed (commented):
Two tiny remaining things, I'm happy for you to squash 👍

nilsdeppe commented:
Thanks for the super quick feedback @fmahebert! Should be all fixed and squashed :)

wthrowe commented:
Travis is complaining, otherwise looks good.
